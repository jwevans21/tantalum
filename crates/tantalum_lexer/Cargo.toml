[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
