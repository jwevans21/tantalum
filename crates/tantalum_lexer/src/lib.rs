//! # Tantalum Lexer
//!
//! This crate provides a lexer for the Tantalum language.

use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxToken};

/// A lexer for the Tantalum language.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Lexer<'file> {
    /// The source file that the lexer is currently lexing.
    file: &'file SourceFile,
    /// The current location in the source file.
    location: SourceLocation,
}

impl<'file> Lexer<'file> {
    /// Create a new lexer from a source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let lexer = Lexer::new(file);
    /// ```
    #[must_use]
    pub fn new(file: &'file SourceFile) -> Self {
        Self {
            file,
            location: file.location_start(),
        }
    }

    /// Get the source file that the lexer is currently lexing.
    #[must_use]
    pub fn source(&self) -> &'file SourceFile {
        self.file
    }

    /// Get the current location in the source file.
    #[must_use]
    pub fn location(&self) -> SourceLocation {
        self.location
    }

    /// Get the character at a location in the source file.
    fn at_location(&self, location: SourceLocation) -> Option<char> {
        self.file.read_char(location)
    }

    /// Extract the next token from the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.next_token().expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    /// ```
    #[must_use]
    pub fn next_token(&mut self) -> Option<SyntaxToken> {
        let start = self.location;

        let (kind, end) = self.characters_to_token(start)?;

        self.location = end;

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Peek at the next token in the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.peek_token(1).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    ///
    /// let token = lexer.peek_token(2).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Whitespace);
    ///
    /// let token = lexer.peek_token(3).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Identifier);
    /// ```
    #[must_use]
    pub fn peek_token(&self, mut n: usize) -> Option<SyntaxToken> {
        let mut start = self.location;

        let (kind, end) = loop {
            n -= 1;
            let (kind, end) = self.characters_to_token(start)?;

            if n == 0 {
                break (kind, end);
            }

            start = end;
        };

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Convert a character to a [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    fn characters_to_token(&self, start: SourceLocation) -> Option<(SyntaxKind, SourceLocation)> {
        let mut current = start;

        let c = self.at_location(current)?;
        current = current.next_by(c.len_utf8());

        let token = match c {
            '(' => SyntaxKind::LParen,
            ')' => SyntaxKind::RParen,
            '{' => SyntaxKind::LBrace,
            '}' => SyntaxKind::RBrace,
            '[' => SyntaxKind::LBracket,
            ']' => SyntaxKind::RBracket,
            '<' => SyntaxKind::LAngle,
            '>' => SyntaxKind::RAngle,
            ',' => SyntaxKind::Comma,
            ':' => SyntaxKind::Colon,
            ';' => SyntaxKind::Semicolon,
            '.' => SyntaxKind::Dot,
            '&' => SyntaxKind::Ampersand,
            '|' => SyntaxKind::Pipe,
            '^' => SyntaxKind::Caret,
            '+' => SyntaxKind::Plus,
            '-' => SyntaxKind::Minus,
            '*' => SyntaxKind::Star,
            '/' => SyntaxKind::Slash,
            '\\' => SyntaxKind::Backslash,
            '%' => SyntaxKind::Percent,
            '=' => SyntaxKind::Equals,
            '!' => SyntaxKind::Bang,
            '~' => SyntaxKind::Tilde,
            '?' => SyntaxKind::Question,
            '@' => SyntaxKind::At,
            '0'..='9' => SyntaxKind::Digit,
            '_' | 'a'..='z' | 'A'..='Z' => {
                while let Some(c) = self.at_location(current) {
                    match c {
                        '_' | 'a'..='z' | 'A'..='Z' | '0'..='9' => {
                            current = current.next_by(c.len_utf8());
                        }
                        _ => break,
                    }
                }

                match self.file.read_span(SourceSpan::new(
                    self.file.id(),
                    self.location.offset(),
                    current.offset(),
                ))? {
                    "fn" => SyntaxKind::Fn,
                    "let" => SyntaxKind::Let,
                    "return" => SyntaxKind::Return,
                    "if" => SyntaxKind::If,
                    "else" => SyntaxKind::Else,
                    "true" => SyntaxKind::True,
                    "false" => SyntaxKind::False,
                    _ => SyntaxKind::Identifier,
                }
            }
            c if c.is_whitespace() => SyntaxKind::Whitespace,
            _ => SyntaxKind::Other,
        };

        Some((token, current))
    }
}

#[test]
fn keywords() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "fn let return if else true false".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 7);

    assert_eq!(tokens[0].kind(), SyntaxKind::Fn);
    assert_eq!(tokens[1].kind(), SyntaxKind::Let);
    assert_eq!(tokens[2].kind(), SyntaxKind::Return);
    assert_eq!(tokens[3].kind(), SyntaxKind::If);
    assert_eq!(tokens[4].kind(), SyntaxKind::Else);
    assert_eq!(tokens[5].kind(), SyntaxKind::True);
    assert_eq!(tokens[6].kind(), SyntaxKind::False);
}

#[test]
fn identifiers() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "a A _ a1 A1 _1 _a _A __AA01".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 9);

    assert_eq!(tokens[0].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[1].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[2].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[3].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[4].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[5].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[6].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[7].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[8].kind(), SyntaxKind::Identifier);
}
