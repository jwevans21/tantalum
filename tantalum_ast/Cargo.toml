[package]
name = "tantalum_ast"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[lints]
workspace = true

