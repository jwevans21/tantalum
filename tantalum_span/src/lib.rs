//! # Tantalum Span
//!
//! Provides a span to locate the positions and ranges of tokens in a file.

use core::{
    fmt::{Debug, Display, Formatter, Result as FmtResult},
    num::Saturating,
    ops::{Deref, Range},
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Location<'file_name> {
    /// The name of the file that this location is in
    #[cfg_attr(feature = "serde", serde(borrow))]
    file_name: &'file_name str,
    /// The byte of the location in the file
    position: Saturating<usize>,
    /// The line number of the location in the file
    line: Saturating<usize>,
    /// The column number of the location in the file
    column: Saturating<usize>,
}

impl<'file_name> Location<'file_name> {
    #[must_use]
    #[inline]
    pub fn new(file_name: &'file_name str) -> Self {
        return Self {
            file_name,
            position: Saturating(0),
            line: Saturating(1),
            column: Saturating(1),
        };
    }

    #[must_use]
    #[inline]
    pub fn new_at(file_name: &'file_name str, position: usize, line: usize, column: usize) -> Self {
        return Self {
            file_name,
            position: Saturating(position),
            line: Saturating(line),
            column: Saturating(column),
        };
    }

    #[inline]
    pub fn advance(&mut self, character: char) {
        if character == '\n' {
            self.line += 1;
            self.column = Saturating(1);
        } else {
            self.column += 1;
        }

        self.position += character.len_utf8();
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.file_name;
    }

    #[must_use]
    #[inline]
    pub fn position(&self) -> usize {
        return self.position.0;
    }

    #[must_use]
    #[inline]
    pub fn line(&self) -> usize {
        return self.line.0;
    }

    #[must_use]
    #[inline]
    pub fn column(&self) -> usize {
        return self.column.0;
    }

    #[must_use]
    #[inline]
    pub fn range(&self, other: &Self) -> Range<usize> {
        return (self.position())..(other.position());
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Span<'file_name> {
    /// The starting location of the span
    #[cfg_attr(feature = "serde", serde(borrow))]
    start: Location<'file_name>,
    /// The ending location of the span
    end: Location<'file_name>,
}

impl<'file_name> Span<'file_name> {
    #[must_use]
    #[inline]
    pub fn new(start: Location<'file_name>, end: Location<'file_name>) -> Self {
        return Self { start, end };
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.start.file_name();
    }

    #[must_use]
    #[inline]
    pub fn range(&self) -> Range<usize> {
        return self.start.range(&self.end);
    }

    #[must_use]
    #[inline]
    pub fn start(&self) -> Location<'file_name> {
        return self.start;
    }

    #[must_use]
    #[inline]
    pub fn end(&self) -> Location<'file_name> {
        return self.end;
    }

    #[must_use]
    #[inline]
    pub fn line(&self) -> usize {
        return self.start.line();
    }

    #[must_use]
    #[inline]
    pub fn column(&self) -> usize {
        return self.start.column();
    }
}

impl Display for Span<'_> {
    #[inline]
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        return write!(f, "{}..{}", self.start.position(), self.end.position());
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Spanned<'file_name, T: Debug + Clone + PartialEq + Eq> {
    #[cfg_attr(feature = "serde", serde(borrow))]
    span: Span<'file_name>,
    data: T,
}

impl<'file_name, T> Spanned<'file_name, T>
where
    T: Debug + Clone + PartialEq + Eq,
{
    #[must_use]
    #[inline]
    pub fn new(span: Span<'file_name>, data: T) -> Self {
        return Self { span, data };
    }

    #[must_use]
    #[inline]
    pub fn spanning(start: Location<'file_name>, end: Location<'file_name>, data: T) -> Self {
        return Self::new(Span::new(start, end), data);
    }

    #[must_use]
    #[inline]
    pub fn join_spans(start: Span<'file_name>, end: Span<'file_name>, data: T) -> Self {
        return Self::new(Span::new(start.start, end.end), data);
    }

    pub fn map<U, F>(self, f: F) -> Spanned<'file_name, U>
    where
        U: Debug + Clone + PartialEq + Eq,
        F: FnOnce(T) -> U,
    {
        return Spanned::new(self.span, f(self.data));
    }

    #[must_use]
    #[inline]
    pub fn start(&self) -> Location<'file_name> {
        return self.span.start();
    }

    #[must_use]
    #[inline]
    pub fn end(&self) -> Location<'file_name> {
        return self.span.end();
    }

    #[must_use]
    #[inline]
    pub fn data(&self) -> &T {
        return &self.data;
    }

    #[must_use]
    #[inline]
    pub fn span(&self) -> Span<'file_name> {
        return self.span;
    }

    #[must_use]
    #[inline]
    pub fn range(&self) -> Range<usize> {
        return self.span.range();
    }

    #[must_use]
    #[inline]
    pub fn line(&self) -> usize {
        return self.span.line();
    }

    #[must_use]
    #[inline]
    pub fn column(&self) -> usize {
        return self.span.column();
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.span.file_name();
    }
}

impl<T> Copy for Spanned<'_, T> where T: Copy + Debug + Clone + PartialEq + Eq {}

impl<T> AsRef<T> for Spanned<'_, T>
where
    T: Debug + Clone + PartialEq + Eq,
{
    #[inline]
    fn as_ref(&self) -> &T {
        return &self.data;
    }
}

impl<T> Deref for Spanned<'_, T>
where
    T: Debug + Clone + PartialEq + Eq,
{
    type Target = T;

    #[inline]
    fn deref(&self) -> &Self::Target {
        return &self.data;
    }
}

impl<T> Display for Spanned<'_, T>
where
    T: Display + Debug + Clone + PartialEq + Eq,
{
    #[inline]
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        return write!(f, "{} @ {}", self.data, self.span);
    }
}
