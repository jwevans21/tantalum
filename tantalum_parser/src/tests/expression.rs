use tantalum_ast::{BinaryOperator, BinaryOperatorKind, Expression, ExpressionKind, Type, TypeKind};
use tantalum_lexer::Lexer;
use tantalum_span::Span;

use pretty_assertions::assert_eq;

use crate::Parser;

macro_rules! prefix_unary_expression_test_cases {
    {
        $($kind:ident: $operator:literal),*
    } => {
        $(
            #[test]
            #[allow(non_snake_case)]
            fn $kind() {
                const SOURCE: &str = concat!($operator, "1");
                let lexer = tantalum_lexer::Lexer::new(stringify!($kind), SOURCE);
                let mut parser = crate::Parser::new(lexer);

                let result = parser.parse_expression();

                pretty_assertions::assert_eq!(
                    result,
                    Ok(
                        tantalum_ast::Expression {
                            span: tantalum_span::Span::new(stringify!($kind), 0, SOURCE.len(), 1, 1),
                            kind: tantalum_ast::ExpressionKind::UnaryOperation {
                                operator: tantalum_ast::UnaryOperator {
                                    span: tantalum_span::Span::new(stringify!($kind), 0, 1, 1, 1),
                                    kind: tantalum_ast::UnaryOperatorKind::$kind
                                },
                                operand: Box::new(
                                    tantalum_ast::Expression {
                                        span: tantalum_span::Span::new(stringify!($kind), 1, SOURCE.len(), 1, 2),
                                        kind: tantalum_ast::ExpressionKind::IntegerLiteral {
                                            value: "1",
                                            radix: 10
                                        }
                                    }
                                )
                            }
                        }
                    )
                );
            }
        )*
    };
}

macro_rules! postfix_unary_expression_test_cases {
    {
        $($kind:ident: $operator:literal),*
    } => {
        $(
            #[test]
            #[allow(non_snake_case)]
            fn $kind() {
                const SOURCE: &str = concat!("1", $operator);
                let lexer = tantalum_lexer::Lexer::new(stringify!($kind), SOURCE);
                let mut parser = crate::Parser::new(lexer);

                let result = parser.parse_expression();

                pretty_assertions::assert_eq!(
                    result,
                    Ok(
                        tantalum_ast::Expression {
                            span: tantalum_span::Span::new(stringify!($kind), 0, SOURCE.len(), 1, 1),
                            kind: tantalum_ast::ExpressionKind::UnaryOperation {
                                operator: tantalum_ast::UnaryOperator {
                                    span: tantalum_span::Span::new(stringify!($kind), 1, SOURCE.len(), 1, 2),
                                    kind: tantalum_ast::UnaryOperatorKind::$kind
                                },
                                operand: Box::new(
                                    tantalum_ast::Expression {
                                        span: tantalum_span::Span::new(stringify!($kind), 0, 1, 1, 1),
                                        kind: tantalum_ast::ExpressionKind::IntegerLiteral {
                                            value: "1",
                                            radix: 10
                                        }
                                    }
                                )
                            }
                        }
                    )
                );
            }
        )*
    }
}

macro_rules! binary_expression_test_cases {
    {
        $($kind:ident: $operator:literal),*
    } => {
        #[allow(non_snake_case)]
        mod BinaryOperation {
            $(
                #[test]
                #[allow(non_snake_case)]
                fn $kind() {
                    const SOURCE: &str = concat!("1 ", $operator, " 2");
                    let lexer = tantalum_lexer::Lexer::new(stringify!($kind), SOURCE);
                    let mut parser = crate::Parser::new(lexer);

                    let result = parser.parse_expression();

                    pretty_assertions::assert_eq!(
                        result,
                        Ok(
                            tantalum_ast::Expression {
                                span: tantalum_span::Span::new(stringify!($kind), 0, SOURCE.len(), 1, 1),
                                kind: tantalum_ast::ExpressionKind::BinaryOperation {
                                    left: Box::new(
                                        tantalum_ast::Expression {
                                            span: tantalum_span::Span::new(stringify!($kind), 0, 1, 1, 1),
                                            kind: tantalum_ast::ExpressionKind::IntegerLiteral {
                                                value: "1",
                                                radix: 10
                                            }
                                        }
                                    ),
                                    operator: tantalum_ast::BinaryOperator {
                                        span: tantalum_span::Span::new(stringify!($kind), 2, (2 + ($operator).len()), 1, 3),
                                        kind: tantalum_ast::BinaryOperatorKind::$kind
                                    },
                                    right: Box::new(
                                        tantalum_ast::Expression {
                                            span: tantalum_span::Span::new(stringify!($kind), (3 + ($operator).len()), SOURCE.len(), 1, (4 + ($operator).len())),
                                            kind: tantalum_ast::ExpressionKind::IntegerLiteral {
                                                value: "2",
                                                radix: 10
                                            }
                                        }
                                    )
                                }
                            }
                        )
                    );
                }
            )*
        }
    };
}

prefix_unary_expression_test_cases! {
    Negation: "-",
    LogicalNegation: "!",
    BitwiseNegation: "~"
}

postfix_unary_expression_test_cases! {
    AddressOf: ".&",
    Deref: ".*"
}

binary_expression_test_cases! {
    Addition: "+",
    Subtraction: "-",
    Multiplication: "*",
    Division: "/",
    Modulus: "%",

    LogicalAnd: "&&",
    LogicalOr: "||",

    Equality: "==",
    NotEqual: "!=",
    LessThan: "<",
    LessThanOrEqual: "<=",
    GreaterThan: ">",
    GreaterThanOrEqual: ">=",

    BitwiseAnd: "&",
    BitwiseOr: "|",
    BitwiseXor: "^",

    ShiftLeft: "<<",
    ShiftRight: ">>"
}

#[test]
fn function_call() {
    const SOURCE: &str = "foo()";
    let lexer = tantalum_lexer::Lexer::new("function_call", SOURCE);
    let mut parser = crate::Parser::new(lexer);

    let result = parser.parse_expression();

    pretty_assertions::assert_eq!(
        result,
        Ok(tantalum_ast::Expression {
            span: tantalum_span::Span::new("function_call", 0, SOURCE.len(), 1, 1),
            kind: tantalum_ast::ExpressionKind::FunctionCall {
                source: Box::new(tantalum_ast::Expression {
                    span: tantalum_span::Span::new("function_call", 0, 3, 1, 1),
                    kind: tantalum_ast::ExpressionKind::Variable { name: "foo" }
                }),
                arguments: vec![]
            }
        })
    );
}

#[test]
fn function_call_with_arguments() {
    const SOURCE: &str = "foo(1, 2)";
    let lexer = tantalum_lexer::Lexer::new("function_call_with_arguments", SOURCE);
    let mut parser = crate::Parser::new(lexer);

    let result = parser.parse_expression();

    pretty_assertions::assert_eq!(
        result,
        Ok(tantalum_ast::Expression {
            span: tantalum_span::Span::new("function_call_with_arguments", 0, SOURCE.len(), 1, 1),
            kind: tantalum_ast::ExpressionKind::FunctionCall {
                source: Box::new(tantalum_ast::Expression {
                    span: tantalum_span::Span::new("function_call_with_arguments", 0, 3, 1, 1),
                    kind: tantalum_ast::ExpressionKind::Variable { name: "foo" }
                }),
                arguments: vec![
                    tantalum_ast::Expression {
                        span: tantalum_span::Span::new("function_call_with_arguments", 4, 5, 1, 5),
                        kind: tantalum_ast::ExpressionKind::IntegerLiteral {
                            value: "1",
                            radix: 10
                        }
                    },
                    tantalum_ast::Expression {
                        span: tantalum_span::Span::new("function_call_with_arguments", 7, 8, 1, 8),
                        kind: tantalum_ast::ExpressionKind::IntegerLiteral {
                            value: "2",
                            radix: 10
                        }
                    }
                ]
            }
        })
    );
}

#[test]
fn array_access() {
    const SOURCE: &str = "foo[1]";
    let lexer = tantalum_lexer::Lexer::new("array_access", SOURCE);
    let mut parser = crate::Parser::new(lexer);

    let result = parser.parse_expression();

    pretty_assertions::assert_eq!(
        result,
        Ok(tantalum_ast::Expression {
            span: tantalum_span::Span::new("array_access", 0, SOURCE.len(), 1, 1),
            kind: tantalum_ast::ExpressionKind::ArrayAccess {
                source: Box::new(tantalum_ast::Expression {
                    span: tantalum_span::Span::new("array_access", 0, 3, 1, 1),
                    kind: tantalum_ast::ExpressionKind::Variable { name: "foo" }
                }),
                index: Box::new(tantalum_ast::Expression {
                    span: tantalum_span::Span::new("array_access", 4, 5, 1, 5),
                    kind: tantalum_ast::ExpressionKind::IntegerLiteral {
                        value: "1",
                        radix: 10
                    }
                })
            }
        })
    );
}

#[test]
fn basic_addition() {
    let lexer = Lexer::new("basic_addition", "1 + 2");
    let mut parser = Parser::new(lexer);

    let result = parser.parse_expression();

    assert_eq!(
        result,
        Ok(Expression {
            span: Span::new("basic_addition", 0, 5, 1, 1),
            kind: ExpressionKind::BinaryOperation {
                left: Box::new(Expression {
                    span: Span::new("basic_addition", 0, 1, 1, 1),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "1",
                        radix: 10
                    }
                }),
                operator: BinaryOperator {
                    span: Span::new("basic_addition", 2, 3, 1, 3),
                    kind: BinaryOperatorKind::Addition
                },
                right: Box::new(Expression {
                    span: Span::new("basic_addition", 4, 5, 1, 5),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "2",
                        radix: 10
                    }
                })
            }
        })
    );
}

#[test]
fn multiplication_with_addition_rhs() {
    let lexer = Lexer::new("multiplication_with_addition_rhs", "1 * 2 + 3");
    let mut parser = Parser::new(lexer);

    let result = parser.parse_expression();

    assert_eq!(
        result,
        Ok(Expression {
            span: Span::new("multiplication_with_addition_rhs", 0, 9, 1, 1),
            kind: ExpressionKind::BinaryOperation {
                left: Box::new(Expression {
                    span: Span::new("multiplication_with_addition_rhs", 0, 5, 1, 1),
                    kind: ExpressionKind::BinaryOperation {
                        left: Box::new(Expression {
                            span: Span::new("multiplication_with_addition_rhs", 0, 1, 1, 1),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "1",
                                radix: 10
                            }
                        }),
                        operator: BinaryOperator {
                            span: Span::new("multiplication_with_addition_rhs", 2, 3, 1, 3),
                            kind: BinaryOperatorKind::Multiplication
                        },
                        right: Box::new(Expression {
                            span: Span::new("multiplication_with_addition_rhs", 4, 5, 1, 5),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "2",
                                radix: 10
                            }
                        })
                    }
                }),
                operator: BinaryOperator {
                    span: Span::new("multiplication_with_addition_rhs", 6, 7, 1, 7),
                    kind: BinaryOperatorKind::Addition
                },
                right: Box::new(Expression {
                    span: Span::new("multiplication_with_addition_rhs", 8, 9, 1, 9),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "3",
                        radix: 10
                    }
                })
            }
        })
    );
}

#[test]
fn multiplication_with_addition_lhs() {
    let lexer = Lexer::new("multiplication_with_addition_lhs", "1 + 2 * 3");
    let mut parser = Parser::new(lexer);

    let result = parser.parse_expression();

    assert_eq!(
        result,
        Ok(Expression {
            span: Span::new("multiplication_with_addition_lhs", 0, 9, 1, 1),
            kind: ExpressionKind::BinaryOperation {
                left: Box::new(Expression {
                    span: Span::new("multiplication_with_addition_lhs", 0, 1, 1, 1),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "1",
                        radix: 10
                    }
                }),
                operator: BinaryOperator {
                    span: Span::new("multiplication_with_addition_lhs", 2, 3, 1, 3),
                    kind: BinaryOperatorKind::Addition
                },
                right: Box::new(Expression {
                    span: Span::new("multiplication_with_addition_lhs", 4, 9, 1, 5),
                    kind: ExpressionKind::BinaryOperation {
                        left: Box::new(Expression {
                            span: Span::new("multiplication_with_addition_lhs", 4, 5, 1, 5),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "2",
                                radix: 10
                            }
                        }),
                        operator: BinaryOperator {
                            span: Span::new("multiplication_with_addition_lhs", 6, 7, 1, 7),
                            kind: BinaryOperatorKind::Multiplication
                        },
                        right: Box::new(Expression {
                            span: Span::new("multiplication_with_addition_lhs", 8, 9, 1, 9),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "3",
                                radix: 10
                            }
                        })
                    }
                })
            }
        })
    );
}

#[test]
fn type_cast() {
    let lexer = Lexer::new("type_cast", "1:u8");
    let mut parser = Parser::new(lexer);

    let result = parser.parse_expression();

    assert_eq!(
        result,
        Ok(Expression {
            span: Span::new("type_cast", 0, 4, 1, 1),
            kind: ExpressionKind::TypeCast {
                expression: Box::new(Expression {
                    span: Span::new("type_cast", 0, 1, 1, 1),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "1",
                        radix: 10
                    }
                }),
                ty: Type {
                    span: Span::new("type_cast", 2, 4, 1, 3),
                    kind: TypeKind::Named("u8")
                }
            }
        })
    );
}

#[test]
fn type_cast_with_binary_expression() {
    let lexer = Lexer::new("type_cast_with_binary_expression", "1 + 2:u8");
    let mut parser = Parser::new(lexer);

    let result = parser.parse_expression();

    assert_eq!(
        result,
        Ok(Expression {
            span: Span::new("type_cast_with_binary_expression", 0, 8, 1, 1),
            kind: ExpressionKind::BinaryOperation {
                left: Box::new(Expression {
                    span: Span::new("type_cast_with_binary_expression", 0, 1, 1, 1),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "1",
                        radix: 10
                    }
                }),
                operator: BinaryOperator {
                    span: Span::new("type_cast_with_binary_expression", 2, 3, 1, 3),
                    kind: BinaryOperatorKind::Addition
                },
                right: Box::new(Expression {
                    span: Span::new("type_cast_with_binary_expression", 4, 8, 1, 5),
                    kind: ExpressionKind::TypeCast {
                        ty: Type {
                            span: Span::new("type_cast_with_binary_expression", 6, 8, 1, 7),
                            kind: TypeKind::Named("u8")
                        },
                        expression: Box::new(Expression {
                            span: Span::new("type_cast_with_binary_expression", 4, 5, 1, 5),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "2",
                                radix: 10
                            }
                        })
                    }
                })
            }
        })
    );
}
