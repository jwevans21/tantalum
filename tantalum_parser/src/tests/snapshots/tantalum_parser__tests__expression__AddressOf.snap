---
source: tantalum_parser/src/tests/expression.rs
expression: result
snapshot_kind: text
---
Ok(Expression(
  span: Span(
    start: Location(
      file_name: "AddressOf",
      position: 0,
      line: 1,
      column: 1,
    ),
    end: Location(
      file_name: "AddressOf",
      position: 3,
      line: 1,
      column: 4,
    ),
  ),
  kind: UnaryOperation(
    operator: UnaryOperator(
      span: Span(
        start: Location(
          file_name: "AddressOf",
          position: 1,
          line: 1,
          column: 2,
        ),
        end: Location(
          file_name: "AddressOf",
          position: 3,
          line: 1,
          column: 4,
        ),
      ),
      kind: AddressOf,
    ),
    operand: Expression(
      span: Span(
        start: Location(
          file_name: "AddressOf",
          position: 0,
          line: 1,
          column: 1,
        ),
        end: Location(
          file_name: "AddressOf",
          position: 1,
          line: 1,
          column: 2,
        ),
      ),
      kind: IntegerLiteral(
        value: "1",
        radix: 10,
      ),
    ),
  ),
))
